### PR TITLE
Test Cache::toJson throws exception on invalid data

### PR DESCRIPTION
**Pull request by ntzm**

### CODE REVIEW

File: src/Cache/Cache.php [hunk 1]
```diff
@@ -79,7 +79,7 @@ public function toJson()
             'hashes' => $this->hashes,
         ));
 
-        if (false === $json) {
+        if (null === $json || false === $json) {
```
Comment (julienfalque) [inline src/Cache/Cache.php]: Is `json_encode()` really returning `null` in case of failure?

File: tests/Cache/CacheTest.php [hunk 2]
```diff
@@ -194,6 +194,25 @@ public function provideCanConvertToAndFromJsonCases()
...
+    {
+        $this->setExpectedException(
+            'UnexpectedValueException',
+            'Can not encode cache signature to JSON, error: "Malformed UTF-8 characters, possibly incorrectly encoded". If you have non-UTF8 chars in your signature, like in license for `header_comment`, consider enabling `ext-mbstring` or install `symfony/polyfill-mbstring`.'
+        );
+
+        $invalidUtf8Sequence = "\xB1\x31";
+
+        $signature = $this->prophesize('PhpCsFixer\Cache\SignatureInterface');
+        $signature->getPhpVersion()->willReturn($invalidUtf8Sequence);
```
Comment (julienfalque) [inline tests/Cache/CacheTest.php]: I would rather use the invalid UTF-8 sequence in `getRules()` as it is more likely to return such value.

File: src/Cache/Cache.php (hunk 1 above)
Comment (ntzm) [reply src/Cache/Cache.php]: Didn't read the docs correctly, updated for a better solution

File: tests/Cache/CacheTest.php [hunk 3]
```diff
@@ -194,6 +194,27 @@ public function provideCanConvertToAndFromJsonCases()
         );
     }
 
+    public function testToJsonThrowsExceptionOnInvalid()
+    {
+        $this->setExpectedException(
```
Comment (julienfalque) [inline tests/Cache/CacheTest.php]: Please set this expectation right before calling `toJson()`.

**Comment by keradus** [comment]
Thanks @ntzm !
Merged manually into #3723

**Pull request closed without merging**